src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,525 @@ impl Display for SnapKey {
     }
 }
 
-/// Name prefix for the self-generated snapshot file.
-const SNAP_GEN_PREFIX: &'static str = "gen";
-/// Name prefix for the received snapshot file.
-const SNAP_REV_PREFIX: &'static str = "rev";
+pub trait SendSnapshotFileCreator {
+    fn create(&mut self, snap: &DbSnapshot, region: &Region) -> raft::Result<()>;
+    fn total_size(&self) -> io::Result<u64>;
+}
 
-/// A structure represents the snapshot file.
-///
-/// All changes to the file will be written to `tmp_file` first, and use
-/// `save` method to make them persistent. When saving a crc32 checksum
-/// will be appended to the file end automatically.
-pub struct SnapFile {
-    file: PathBuf,
-    digest: Digest,
-    size_track: Arc<RwLock<u64>>,
-    // File is the file obj represent the tmpfile, string is the actual path to
-    // tmpfile.
-    tmp_file: Option<(File, String)>,
+pub trait SnapshotFileCommon {
+    fn path(&self) -> &str;
+    fn exists(&self) -> bool;
 }
 
-impl SnapFile {
-    fn new<T: Into<PathBuf>>(snap_dir: T,
-                             size_track: Arc<RwLock<u64>>,
-                             is_sending: bool,
-                             key: &SnapKey)
-                             -> io::Result<SnapFile> {
-        let mut file_path = snap_dir.into();
-        if !file_path.exists() {
-            try!(fs::create_dir_all(file_path.as_path()));
-        }
-        let prefix = if is_sending {
-            SNAP_GEN_PREFIX
-        } else {
-            SNAP_REV_PREFIX
-        };
-        let file_name = format!("{}_{}.snap", prefix, key);
-        file_path.push(&file_name);
-
-        let mut f = SnapFile {
-            file: file_path,
-            digest: Digest::new(crc32::IEEE),
-            size_track: size_track,
-            tmp_file: None,
-        };
-        try!(f.init());
-        Ok(f)
-    }
-
-    pub fn init(&mut self) -> io::Result<()> {
-        if self.exists() || self.tmp_file.is_some() {
-            return Ok(());
-        }
+pub trait SendSnapshotFile: SnapshotFileCommon + Read {
+    fn meta(&self) -> io::Result<Metadata>;
+    fn total_size(&self) -> io::Result<u64>;
+    fn delete(&self);
+}
 
-        let tmp_path = format!("{}{}", self.path().display(), TMP_FILE_SUFFIX);
-        let tmp_f = try!(OpenOptions::new().write(true).create_new(true).open(&tmp_path));
-        self.tmp_file = Some((tmp_f, tmp_path));
-        Ok(())
-    }
+pub trait RecvSnapshotFileWriter: SnapshotFileCommon + Write + Send {
+    fn save(&mut self) -> io::Result<()>;
+}
 
-    pub fn meta(&self) -> io::Result<Metadata> {
-        self.file.metadata()
-    }
+pub struct ApplyOptions<'a> {
+    pub db: Arc<DB>,
+    pub region: &'a Region,
+    pub abort: Arc<AtomicUsize>,
+    pub write_batch_size: usize,
+    pub snapshot_size: &'a mut usize,
+    pub snapshot_kv_count: &'a mut usize,
+}
+
+pub trait RecvSnapshotFileApplier: SnapshotFileCommon {
```

Comment:
What's the difference between `RecvSnapshotFileWriter` and `RecvSnapshotFileApplier`? I think it will be more clear that Applier has only one method named `apply`, and accept a `RecvSnapshotFile` as args.